app/services/publish_service.rb

```diff
@@ -1,11 +1,12 @@
 # frozen_string_literal: true
 
 class PublishService
-  attr_reader :document
-  delegate :current_edition, :live_edition, to: :document
+  attr_reader :document, :current_edition
+  delegate :live_edition, to: :document
 
-  def initialize(document)
-    @document = document
+  def initialize(edition)
+    @current_edition = edition
+    @document = edition.document
```

Comment:
I feel like the changes here have impacted this class negatively. It seems a bit strange we have instance variables for document and current_edition but not live edition.

Was there a particular reason this was changed?

If so I wonder if it'd be cleaner if we had an instance variable of edition and delegated document to that.

Reply:
So this is because of the ambiguous `inverse_of` between current/live editions and their document. We're passing `edition.document` to the service, but this doesn't prevent a DB query when the service tries to do `(edition.)document.current_edition`. The end result is that the `edition` variable in the controller needs to be reloaded.

It looks like this is actually 'fixed' automatically by the serialisation/deserialisation that happens when we pass the edition to the mailer and invoke `deliver_later`. So we could just roll with that, but it's not very intuitive. What might be better is to run the controller by finding/locking on the document and using `with_current_edition`.

How does that sound?

Reply:
Oh darn, that makes sense. Yeah that suggestion sounds good - looks like we're only really dealing with a document in the controller anyway.

Reply:
Turns out that doesn't work either, as the PublishService is modifying the attributes on the individual editions, rather than re-assigning stuff on the document. I had a go at changing the service, but the assignment for `live_edition` doesn't work, as re-assigning it tries to set `document_id` to nil on the previous live edition, which breaks a foreign key constraint.

So now we have something else, which actually points to some kind of global `context` variable :-0.

Reply:
Refactored again as discussed :-).